web3c/invoke_subscriptions.js

```diff
@@ -172,10 +208,21 @@ class InvokeSubscriptions {
     if (subscription.expectedTransactions[data.transactionHash]) {
       const expectedTransaction = subscription.expectedTransactions[data.transactionHash];
       delete subscription.expectedTransactions[data.transactionHash];
+      clearTimeout(expectedTransaction.timeout);
+      expectedTransaction.timeout = null;
       this.forwardData(data, expectedTransaction.toAddress, expectedTransaction.promise);
 
     } else {
-      subscription.receivedTransactions[data.transactionHash] = data;
+      subscription.receivedTransactions[data.transactionHash] = {
+        transactionHash: data.transactionHash,
+        returnData: data.returnData,
+        timeout: null
+      };
+
+      subscription.receivedTransactions[data.transactionHash].timeout = setTimeout(() => {
```

Comment:
can we have the setTimeout and associated lambda function defined once? it seems to be set with slightly different semantics here and on line 145

Reply:
The problem is how these setTimeouts are set:

 * `subscription.receivedTransactions[data.transactionHash].timeout` is set when an event is received by the subscription. The user may be interested in this event or not. If not, we may want to cleanup after a while to avoid having that structure growing indefinitely

 * `subscription.expectedTransactions[...].timeout` is set when the user expresses interest in a specific transaction. If that transaction is not yet available, the user will need to wait.

Since both operations are asynchronous, and we don't know if any of them will occur, we need both timeouts.

If we want to avoid setting all these timeouts alternatives that I can think of are:

 * no cleanup timeouts
 * having a global cleanup function that goes through all the subscriptions every time and cleans up the subscriptions that have not been active for a while

The second option may be easier to understand, it will just go through all the subscriptions every time, but if the number of subscriptions is small, this should not be a problem.

However, timing out failed requests will not be as accurate.